torch/jit/_script.py

```diff
@@ -270,6 +269,154 @@ def __getattr__(self, attr):
 
 
 if _enabled:
+
+    class ScriptObjectWrapper(object):
+        """
+        An analogue of RecursiveScriptModule for regular objects that are not modules.
+        This class is a wrapper around a torch._C.ScriptObject that represents an instance
+        of a TorchScript class and allows it to be used in Python.
+        Attributes:
+            _c [torch._C.ScriptObject]: The C++ object to which attribute lookups and method
+                calls are forwarded.
+            _props [Dict[str, property]]: A dictionary of properties fetched from self._c and
+                exposed on this wrppaer.
+        """
+        def __init__(self, cpp_class):
+            super(ScriptObjectWrapper, self).__init__()
+            self.__dict__["_initializing"] = True
+            self._c = cpp_class
+            # Add wrapped object's properties to this class instance.
+            self._props = {prop[0]: property(prop[1], prop[2]) for prop in self._c._properties()}
+            self.__dict__["_initializing"] = False
+
+        def __getattr__(self, attr):
+            if "_initializing" in self.__dict__ and self.__dict__["_initializing"]:
+                return super(ScriptObjectWrapper, self).__getattr__(attr)  # type: ignore
+            if attr in self._props:
+                return self._props[attr].fget()
+            return getattr(self._c, attr)
+
+        def __setattr__(self, attr, value):
+            if "_initializing" in self.__dict__ and self.__dict__["_initializing"]:
+                return super(ScriptObjectWrapper, self).__setattr__(attr, value)
+            if attr in self._props:
+                self._props[attr].fset(value)
+                return
+            setattr(self._c, attr, value)
+
+        # Delegate calls to magic methods like __len__ to the C++ module backing the
+        # ScriptObjectWrapper.
+        def forward_magic_method(self, method_name, *args, **kwargs):
+            if not self._c._has_method(method_name):
+                raise TypeError()
+
+            self_method = self.__getattr__(method_name)
+            return self_method(*args, **kwargs)
+
+        def __getstate__(self):
+            return self.forward_magic_method("__getstate__")
+
+        def __setstate__(self, val):
+            self.forward_magic_method("__setstate__", val)
+
+        def __iter__(self):
```

Comment:
We probably don't need all of these `__` methods for Torchbind. I think we will expand this list later on, but for now I think we should be able to get by with just `__getstate__` and `__setstate__` and maybe a few others.

Reply:
ok